proc/threads.c

```diff
@@ -1596,6 +1596,10 @@ static int _proc_lockSet(lock_t *lock, int interruptible, spinlock_ctx_t *scp)
 		do {
 			/* _proc_lockUnlock will give us a lock by it's own */
 			if (proc_threadWaitEx(&lock->queue, &lock->spinlock, 0, interruptible, scp) == -EINTR) {
+				/* Can happen when thread_destroy is called on lock owner and current */
+				if (lock->owner == NULL) {
+					return -EINTR;
+				}
```

Comment:
IMHO the core of the problem is that we clear the `threads_common.spinlock` in line 1594, which can lead to a race on a SMP system.

So I think this check should be performed under `threads_common.spinlock` in the line 1606 instead. Current solution just makes the race less probable

Reply:
Eh, I guess the whole `if (lock->owner != current)` should be under spinlock too, I think